File: dotnet/src/Core/Broker.cs
using GuerrillaNtp;
using Microsoft.IdentityModel.Tokens;
using MQTTnet;
using MQTTnet.Client;
using MQTTnet.Protocol;
using MQTTnet.Formatter;
using MQTTnet.Diagnostics;
using Timer = System.Timers.Timer;
using System.Text.Json;
using Microsoft.Extensions.Logging;
using Agience.Core.Models.Messages;

namespace Agience.Core
{
    public class Broker
    {
        internal event Func<Task> Disconnected
        {
            add => _mqttClient.DisconnectedAsync += async (args) => await value();
            remove => _mqttClient.DisconnectedAsync -= async (args) => await value();
        }

        NtpClient? _ntpClient;

        string? _customNtpHost = null;

        //https://www.ntppool.org/zone/@
        List<string> ntpHosts = new() {
            "pool.ntp.org",
            "north-america.pool.ntp.org",
            "europe.pool.ntp.org",
            "asia.pool.ntp.org",
            "south-america.pool.ntp.org",
            "africa.pool.ntp.org",
            "oceania.pool.ntp.org"};

        public string Timestamp => (_ntpClient?.Last ?? throw new InvalidOperationException()).Now.UtcDateTime.ToString(TIME_FORMAT);

        private Timer _ntpTimer = new Timer(TimeSpan.FromDays(1).TotalMilliseconds); // Synchronize daily

        public bool IsConnected => _mqttClient.IsConnected;

        private const string MESSAGE_TYPE_KEY = "message.type";
        private const string TIME_FORMAT = "yyyy-MM-ddTHH:mm:ss.fff";

        private readonly IMqttClient _mqttClient;
        private readonly Dictionary<string, List<CallbackContainer>> _callbacks = new();

        private readonly ILogger<Broker> _logger;        

        internal Broker(ILogger<Broker> logger, string? customNtpHost = null)
        {
            _logger = logger;
            _customNtpHost = customNtpHost;
            _mqttClient = new MqttFactory().CreateMqttClient(new MqttNetLogger() { IsEnabled = true });
            _mqttClient.ApplicationMessageReceivedAsync += _client_ApplicationMessageReceivedAsync;
        }

        internal async Task Connect(string token, string brokerUri)
        {
            await StartNtpClock();

            _logger.LogInformation($"Broker Connected Status: {IsConnected}");

            if (!_mqttClient.IsConnected)
            {
                _logger.LogInformation($"Connecting to {brokerUri}");

                var options = new MqttClientOptionsBuilder()
                    .WithWebSocketServer(configure => { configure.WithUri(brokerUri); })
                    .WithTlsOptions(configure => { configure.UseTls(true); })
                    .WithCredentials(token, "<no_password>")
                    .WithProtocolVersion(MqttProtocolVersion.V500)
                    .WithoutThrowOnNonSuccessfulConnectResponse()
                    .WithTimeout(TimeSpan.FromSeconds(300))
                    .WithKeepAlivePeriod(TimeSpan.FromSeconds(60))
                    .WithSessionExpiryInterval(60)
                    .WithCleanStart()
                    .Build();

                await _mqttClient.ConnectAsync(options);

                if (_mqttClient.IsConnected)
                {
                    _logger.LogInformation($"Broker Connected");
                }
                else
                {
                    _logger.LogInformation($"Broker Connection Failed");
                }                
            }
        }

        private Task _client_ApplicationMessageReceivedAsync(MqttApplicationMessageReceivedEventArgs args)
        {
            var topic = args.ApplicationMessage.Topic;
            var callbackTopic = topic.Substring(topic.IndexOf('/') + 1); // Remove the SenderId segment

            if (_callbacks.TryGetValue(callbackTopic, out var callbackContainers))
            {
                var messageType = Enum.TryParse<BrokerMessageType>(
                    args.ApplicationMessage.UserProperties.FirstOrDefault(x => x.Name == MESSAGE_TYPE_KEY)?.Value.ToString(), out var parsedMessageType) ?
                    parsedMessageType :
                    BrokerMessageType.UNKNOWN;

                var message = new BrokerMessage()
                {
                    Type = messageType,
                    Topic = args.ApplicationMessage.Topic //,
                    //Payload = args.ApplicationMessage.ConvertPayloadToString()
                };

                switch (messageType)
                {
                    case BrokerMessageType.EVENT:
                        message.Data = args.ApplicationMessage.ConvertPayloadToString();
                        break;
                    case BrokerMessageType.INFORMATION:
                        var payloadString = args.ApplicationMessage.ConvertPayloadToString();
                        message.Information = JsonSerializer.Deserialize<Information>(payloadString);
                        break;
                }

                foreach (var container in callbackContainers)
                {
                    if (container.Callback != null)
                    {
                        Task.Run(async () =>
                        {
                            try
                            {
                                await Task.Run(() => container.Callback(message));
                            }
                            catch (Exception ex)
                            {
                                _logger.LogError(ex, "An error occurred in the callback.");
                                throw;
                            }
                        });
                    }
                }
            }
            return Task.CompletedTask;
        }

        internal async Task Subscribe(string topic, Func<BrokerMessage, Task> callback)
        {
            if (!_mqttClient.IsConnected) throw new InvalidOperationException("Not Connected");

            var callbackTopic = topic.Substring(topic.IndexOf('/') + 1); // Remove the SenderId segment

            var container = new CallbackContainer(callback);
            if (!_callbacks.ContainsKey(callbackTopic))
            {
                _callbacks[callbackTopic] = new List<CallbackContainer>();
            }
            _callbacks[callbackTopic].Add(container);

            var options = new MqttClientSubscribeOptionsBuilder()
                .WithTopicFilter(topic, MqttQualityOfServiceLevel.AtMostOnce)
                .Build();

            await _mqttClient.SubscribeAsync(options);
        }

        internal async Task Disconnect()
        {
            if (_mqttClient.IsConnected)
            {
                await _mqttClient.TryDisconnectAsync();
            }
        }

        internal async Task Unsubscribe(string topic)
        {
            var callbackTopic = topic.Substring(topic.IndexOf('/') + 1);

            _callbacks.Remove(callbackTopic);

            await _mqttClient.UnsubscribeAsync(callbackTopic);
        }

        internal void Publish(BrokerMessage message)
        {
            PublishAsync(message).ContinueWith(task =>
            {
                if (task.IsFaulted && task.Exception != null)
                {
                    throw task.Exception;
                }
            }, TaskScheduler.Current);
        }

        internal async Task PublishAsync(BrokerMessage message)
        {
            if (_mqttClient.IsConnected)
            {
                string payload = string.Empty;

                switch (message.Type)
                {
                    case BrokerMessageType.EVENT:
                        payload = message.Data?.ToString() ?? string.Empty;
                        break;
                    case BrokerMessageType.INFORMATION:
                        payload = JsonSerializer.Serialize(message.Information);
                        break;
                }

                var mqMessage = new MqttApplicationMessageBuilder()
                    .WithTopic(message.Topic ?? throw new ArgumentNullException(nameof(message.Topic)))
                    .WithPayload(payload)
                    .WithRetainFlag(false)
                    .WithUserProperty(MESSAGE_TYPE_KEY, message.Type.ToString())
                    .WithQualityOfServiceLevel(MqttQualityOfServiceLevel.AtMostOnce)
                    .Build();

                await _mqttClient.PublishAsync(mqMessage);
            }
        }

        private class MqttNetLogger : IMqttNetLogger
        {
            public bool IsEnabled { get; internal set; }

            public void Publish(MqttNetLogLevel logLevel, string source, string message, object[] parameters, Exception exception)
            {
                // TODO: Write to real logger
                // Console.WriteLine($"{logLevel}: {source} - {message}");
            }
        }

        private async Task StartNtpClock()
        {
            await QueryNtpWithBackoff(); // Query now

            _ntpTimer.Elapsed += async (sender, args) => { await QueryNtpWithBackoff(); };
            _ntpTimer.AutoReset = true;
            _ntpTimer.Start();
        }

        private async Task QueryNtpWithBackoff(double maxDelaySeconds = 32)
        {
            //Using a custom host from the settings, instead of the pre-defined list.
            if (!_customNtpHost.IsNullOrEmpty())
            {

                if (!string.IsNullOrEmpty(_customNtpHost) && !_customNtpHost.ToLower().EndsWith("pool.ntp.org"))
                    throw new ArgumentException("The CustomNtpHost must end with `pool.ntp.org`.");

                ntpHosts.Clear();
                ntpHosts.Add(_customNtpHost);
            }

            var delay = TimeSpan.FromSeconds(1);
            var currentNtpHostIndex = 1;
            while (true)
            {
                var ntpHpst = ntpHosts[currentNtpHostIndex - 1];
                try
                {
                    _ntpClient = new(ntpHpst);
                    _logger.LogInformation($"NTP Querying host {ntpHpst}");
                    _ntpClient.Query();
                    _logger.LogInformation($"Connected to {ntpHpst}. NTP Time: {Timestamp}");
                    break;
                }
                catch (Exception ex)
                {
                    _logger.LogError($"NTP Query to host {ntpHpst} failed");

                    var startNewCycle = currentNtpHostIndex == ntpHosts.Count();

                    currentNtpHostIndex = startNewCycle ? 1 : currentNtpHostIndex + 1;

                    if (startNewCycle)
                    {
                        _logger.LogInformation($"Trying again a NTP connection in {delay.TotalSeconds} seconds.\r\n{ex.Message}");
                        await Task.Delay(delay);
                        delay = TimeSpan.FromSeconds(Math.Min(delay.TotalSeconds * 2, maxDelaySeconds));
                    }
                }
            }
        }

    }

    internal class CallbackContainer
    {
        //public Guid Id { get; } = Guid.NewGuid(); // Unique identifier for the callback
        public Func<BrokerMessage, Task> Callback { get; set; }
        public CallbackContainer(Func<BrokerMessage, Task> callback)
        {
            Callback = callback;
        }
    }
}

File: python/src/core/broker.py
import ssl
from datetime import datetime
from typing import Dict, List, Optional, Callable, Any
import asyncio
from asyncio import Task
import logging
from ntplib import NTPClient
from pydantic import BaseModel
import paho.mqtt.client as mqtt
from urllib.parse import urlparse

from core.models.messages.broker_message import BrokerMessage, BrokerMessageType
from core.information import Information


class CallbackContainer(BaseModel):
    callback: Callable[[BrokerMessage], Task]


class Broker:
    MESSAGE_TYPE_KEY = "message.type"
    TIME_FORMAT = "%Y-%m-%dT%H:%M:%S.%f"

    def __init__(self, logger: logging.Logger, custom_ntp_host: Optional[str] = None):
        self._logger = logger
        self._custom_ntp_host = custom_ntp_host
        self._mqtt_client = mqtt.Client(protocol=mqtt.MQTTv5)
        self._callbacks: Dict[str, List[CallbackContainer]] = {}
        self._ntp_client: Optional[NTPClient] = None
        self._connected = False

        # NTP hosts
        self.ntp_hosts = [
            "pool.ntp.org",
            "north-america.pool.ntp.org",
            "europe.pool.ntp.org",
            "asia.pool.ntp.org",
            "south-america.pool.ntp.org",
            "africa.pool.ntp.org",
            "oceania.pool.ntp.org"
        ]

        # Set up MQTT callbacks

        self._mqtt_client.on_connect = self._on_connect
        self._mqtt_client.on_message = self._on_message
        self._mqtt_client.on_disconnect = self._on_disconnect

    @property
    def is_connected(self) -> bool:
        return self._connected

    @property
    def timestamp(self) -> str:
        if not self._ntp_client:
            raise RuntimeError("NTP client not initialized")
        return datetime.utcfromtimestamp(self._ntp_client.request('pool.ntp.org').tx_time).strftime(self.TIME_FORMAT)

    async def connect(self, token: str, broker_uri: str):
        await self._start_ntp_clock()
        self._logger.info(f"Broker Connected Status: {self.is_connected}")

        if not self.is_connected:
            self._logger.info(f"Connecting to {broker_uri}")

            # Parse the broker URI to get host and port
            # Expected format: "wss://example.com:8883"
            parsed_uri = urlparse(broker_uri)
            host = parsed_uri.hostname
            port = parsed_uri.port or 443  # Default to 443 if not specified

            # Configure MQTT client
            self._mqtt_client.username_pw_set(
                username=token, password="<no_password>")
            self._mqtt_client.transport = "websockets"
            self._mqtt_client.ws_set_options(path="/mqtt")

            # TODO: SSL temp fix
            # self._mqtt_client.tls_set()  # Enable TLS
            self._mqtt_client.tls_set(cert_reqs=ssl.CERT_NONE)
            self._mqtt_client.tls_insecure_set(True)

            try:
                self._mqtt_client.connect(host, port, keepalive=60)
                # self._mqtt_client.loop_start()  # Start network loop in separate thread

                # Start a background loop for the MQTT client
                loop = asyncio.get_running_loop()
                loop.run_in_executor(None, self._mqtt_client.loop_start)

                # Wait for connection or timeout
                timeout = 10
                start_time = datetime.now()
                while not self.is_connected and (datetime.now() - start_time).seconds < timeout:
                    await asyncio.sleep(0.1)

                if not self.is_connected:
                    raise TimeoutError("Connection timeout")

                self._logger.info("Broker Connected")
            except Exception as e:
                self._logger.error(f"Broker Connection Failed: {str(e)}")
                raise

    def _on_connect(self, client, userdata, flags, rc, properties=None):
        if rc == 0:
            self._connected = True
        else:
            self._logger.error(f"Connection failed with code {rc}")

    def _on_message(self, client, userdata, msg: mqtt.MQTTMessage):
        self._logger.info(f"Received Message: {msg.topic}")

        topic_parts = msg.topic.split('/')
        callback_topic = '/'.join(topic_parts[2:])  # Remove SenderId segment

        if callback_topic in self._callbacks:
            try:
                # Get message type from properties
                properties = msg.properties if hasattr(
                    msg, 'properties') else {}
                try:
                    message_type = BrokerMessageType(
                        properties.get(self.MESSAGE_TYPE_KEY, 'UNKNOWN'))
                except ValueError:
                    message_type = BrokerMessageType.UNKNOWN

                message = BrokerMessage(
                    type=message_type,
                    topic=msg.topic
                )

                if message_type == BrokerMessageType.EVENT:
                    message.data = msg.payload.decode()
                elif message_type == BrokerMessageType.INFORMATION:
                    message.information = Information.parse_raw(msg.payload)

                # Create event loop if not exists
                try:
                    loop = asyncio.get_event_loop()
                except RuntimeError:
                    loop = asyncio.new_event_loop()
                    asyncio.set_event_loop(loop)

                for container in self._callbacks[callback_topic]:
                    asyncio.create_task(container.callback(message))

            except Exception as e:
                self._logger.error(f"Message handling error: {str(e)}")
                raise

    def _on_disconnect(self, client, userdata, rc):
        self._connected = False
        self._logger.info(f"Broker disconnected with code: {rc}")

    async def subscribe(self, topic: str, callback: Callable[[BrokerMessage], Task]):
        if not self.is_connected:
            raise RuntimeError("Not Connected")

        topic_parts = topic.split('/')
        callback_topic = '/'.join(topic_parts[2:])

        container = CallbackContainer(callback=callback)

        if callback_topic not in self._callbacks:
            self._callbacks[callback_topic] = []

        self._callbacks[callback_topic].append(container)
        self._mqtt_client.subscribe(topic, qos=0)

    async def unsubscribe(self, topic: str):
        topic_parts = topic.split('/')
        callback_topic = '/'.join(topic_parts[2:])

        if callback_topic in self._callbacks:
            del self._callbacks[callback_topic]

        self._mqtt_client.unsubscribe(topic)

    async def disconnect(self):
        if self.is_connected:
            self._mqtt_client.disconnect()
            self._mqtt_client.loop_stop()
            self._connected = False

    def publish(self, message: BrokerMessage):
        asyncio.create_task(self.publish_async(message))

    async def publish_async(self, message: BrokerMessage):
        if not self.is_connected:
            return

        if not message.topic:
            raise ValueError("Topic cannot be None")

        payload = ""
        if message.type == BrokerMessageType.EVENT:
            payload = str(message.data) if message.data else ""
        elif message.type == BrokerMessageType.INFORMATION:
            payload = message.information.json() if message.information else ""

        properties = mqtt.Properties(
            user_property=(self.MESSAGE_TYPE_KEY, message.type.value)
        )

        self._mqtt_client.publish(
            message.topic,
            payload.encode(),
            qos=0,
            retain=False,
            properties=properties
        )

    async def _start_ntp_clock(self):
        await self._query_ntp_with_backoff()

        # Schedule daily NTP updates
        # while True:
        #     await asyncio.sleep(24 * 60 * 60)  # 24 hours
        #     await self._query_ntp_with_backoff()

    async def _query_ntp_with_backoff(self, max_delay_seconds: float = 32):
        if self._custom_ntp_host:
            if not self._custom_ntp_host.lower().endswith('pool.ntp.org'):
                raise ValueError(
                    "The CustomNtpHost must end with `pool.ntp.org`")
            self.ntp_hosts = [self._custom_ntp_host]

        delay = 1
        current_host_index = 0

        while True:
            ntp_host = self.ntp_hosts[current_host_index]
            try:
                self._ntp_client = NTPClient()
                self._logger.info(f"NTP Querying host {ntp_host}")
                self._ntp_client.request(ntp_host)
                self._logger.info(
                    f"Connected to {ntp_host}. NTP Time: {self.timestamp}")
                break
            except Exception as e:
                self._logger.error(f"NTP Query to host {ntp_host} failed")

                start_new_cycle = current_host_index == len(self.ntp_hosts) - 1
                current_host_index = 0 if start_new_cycle else current_host_index + 1

                if start_new_cycle:
                    self._logger.info(f"Trying again a NTP connection in {
                                      delay} seconds.\n{str(e)}")
                    await asyncio.sleep(delay)
                    delay = min(delay * 2, max_delay_seconds)
